### PR TITLE
PSA: Add support macros for IV/nonce sizes

### LINKED ISSUE #3270: PSA: update support macros for IV/nonce sizes since API 1.0 beta 3
The support macros for cipher IV sizes and AEAD nonce sizes are up to date for PSA Crypto API 1.0 beta 3 but not for 1.0.0 final. The current macros have somewhat inconsistent naming and some are missing or don't take the right parameters.

The goal of this task is to define the new macros. Keep the existing ones as aliases for backward compatibilty in `crypto_compat.h`.

See the [specification](https://armmbed.github.io/mbed-crypto/psa/#application-programming-interface) for a list. The [section on changes](https://armmbed.github.io/mbed-crypto/html/appendix/history.html#changes-between-1-0-beta-3-and-1-0-0) enumerates the macros that have changed since beta 3 (which was aligned with Mbed Crypto by construction). The [private discussion](https://github.com/ARMmbed/psa-crypto-api/pull/293) may be useful.

The following macros are currently missing (they may be present under an older name):
```
PSA_AEAD_NONCE_LENGTH
PSA_AEAD_NONCE_MAX_SIZE
PSA_CIPHER_IV_LENGTH
PSA_CIPHER_IV_MAX_SIZE
```

Related: https://github.com/ARMmbed/mbedtls/issues/3271: the usage of these macros in the library and in tests.

### PR DESCRIPTION
**Pull request by bensze01**
## Description
This commit defines the following new macros added to the PSA Crypto API [between versions 1.0 beta 3 and 1.0.0 final](https://armmbed.github.io/mbed-crypto/html/appendix/history.html#changes-between-1-0-beta-3-and-1-0-0):

[PSA_AEAD_NONCE_LENGTH](https://armmbed.github.io/mbed-crypto/html/api/ops/aead.html#c.PSA_AEAD_NONCE_LENGTH)
[PSA_AEAD_NONCE_MAX_SIZE](https://armmbed.github.io/mbed-crypto/html/api/ops/aead.html#c.PSA_AEAD_NONCE_MAX_SIZE)
[PSA_CIPHER_IV_LENGTH](https://armmbed.github.io/mbed-crypto/html/api/ops/ciphers.html#c.PSA_CIPHER_IV_LENGTH)
[PSA_CIPHER_IV_MAX_SIZE](https://armmbed.github.io/mbed-crypto/html/api/ops/ciphers.html#c.PSA_CIPHER_IV_MAX_SIZE)

The in-code documentation was adapted from the PSA Cryptography API spec (see above the links to the individual macros)

Closes #3270 

## Status
**READY**

## Requires Backporting
NO  

## Migrations
NO

### CODE REVIEW

**Review by yanesca — CHANGES_REQUESTED** [review]
I reviewed the PR and left comments with a suggestion for improvement and a request for change.

File: include/psa/crypto_sizes.h [hunk 1]
```diff
@@ -657,4 +657,95 @@
...
+      (key_type) == PSA_KEY_TYPE_CHACHA20 && \
+          PSA_ALG_AEAD_WITH_DEFAULT_TAG_LENGTH(alg) == PSA_ALG_CHACHA20_POLY1305 ? 12 : \
+     0)
+
+
+/** The maximum nonce size for all supported AEAD algorithms, in bytes.
+ *
+ * See also #PSA_AEAD_NONCE_LENGTH().
+ */
+#define PSA_AEAD_NONCE_MAX_SIZE 12
```
[the remark this answers is not in the document]
Comment (bensze01) [reply include/psa/crypto_sizes.h:695]: @yanesca @gilles-peskine-arm While I thought the wording in the spec as to what PSA_AEAD_NONCE_MAX_SIZE should mean was a bit vague, I took it it mean `MAX(PSA_AEAD_NONCE_LENGTH(key_type, alg), for all key_type, for all alg))`, ie. the maximum of the default nonce lengths for the supported key and algorithm pairs.

I arrived to this conclusion [due to the spec repeatedly suggesting](https://armmbed.github.io/mbed-crypto/html/api/ops/aead.html) one uses either PSA_AEAD_NONCE_LENGTH or PSA_AEAD_NONCE_MAX_SIZE to allocate a buffer large enough to hold the nonce generated by the PSA Crypto implementation, which I took to be the macro's main use case

[the remark this answers is not in the document]
Comment (gilles-peskine-arm) [reply include/psa/crypto_sizes.h:695]: Indeed, the primary intended use for these macros is to determine an output buffer size that's guaranteed to be large enough. This does warrant a clarification in the specification, I'll file it for a 1.0.x update.

[the remark this answers is not in the document]
Comment (yanesca) [reply include/psa/crypto_sizes.h:695]: @gilles-peskine-arm am I interpreting this correctly, do you mean that 12 is the correct value for this macro here?

[the remark this answers is not in the document]
Comment (gilles-peskine-arm) [reply include/psa/crypto_sizes.h:695]: I'm sure that 16 is correct. I'd have to check exactly how we generate nonces for supported algorithms to see if 12 is correct.

[the remark this answers is not in the document]
Comment (yanesca) [reply include/psa/crypto_sizes.h:695]: As far as I can tell we are not generating it yet. We are taking it from the user. `psa_aead_generate_nonce()` does not seem to be implemented yet, but the signature is safe (takes the length of the target buffer). For CCM we need at least 13 to allow all standard options, but other than that it is up to our judgement.

[the remark this answers is not in the document]
Comment (athoelke) [reply include/psa/crypto_sizes.h:695]: > I arrived to this conclusion [due to the spec repeatedly suggesting](https://armmbed.github.io/mbed-crypto/html/api/ops/aead.html) one uses either `PSA_AEAD_NONCE_LENGTH `or `PSA_AEAD_NONCE_MAX_SIZE `to allocate a buffer large enough to hold the nonce generated by the PSA Crypto implementation

👀! - it says this precisely once, in `psa_aead_generate_nonce()`. But the rest of the feedback and @gilles-peskine-arm conclusion is valid.

This is a copy-paste issue arising from the other definitions of `xxx_MAX_SIZE` macros. In this case, this specifically should be the maximum size of **any nonce generated by this function**, not the maximum size of **any AEAD nonce supported by the implementation**.

[the remark this answers is not in the document]
Comment (yanesca) [reply include/psa/crypto_sizes.h:695]: @athoelke Thank you for the clarification! I think I still have a question: what would be the benefit/justification of `psa_aead_generate_nonce()` generating longer nonces than the maximum size of any AEAD nonce supported by the implementation?

[the remark this answers is not in the document]
Comment (athoelke) [reply include/psa/crypto_sizes.h:695]: > what would be the benefit/justification of `psa_aead_generate_nonce()` generating longer nonces than the maximum size of any AEAD nonce supported by the implementation?

That wouldn't make sense. `psa_aead_generate_nonce()` is defined to output `PSA_AEAD_NONE_LENGTH(key_type, alg)`.

@bensze01's came to the correct conclusion that `PSA_AEAD_NONCE_MAX_SIZE` must be a value equal to (or greater than)

max { `PSA_AEAD_NONCE_LENGTH(key_type, alg)`, for all supported AEAD (`key_type`, `alg`) }

`PSA_AEAD_NONCE_LENGTH(key_type, alg)` is not the maximum size of nonce supported as input to `psa_aead_set_nonce()`, `psa_aead_encrypt()` or `psa_aead_decrypt()`, just the default size that is generated by `psa_aead_generate_nonce()`.

[the remark this answers is not in the document]
Comment (bensze01) [reply include/psa/crypto_sizes.h:695]: > 👀! - it says this precisely once, in psa_aead_generate_nonce().

Yeah, sorry. I remembered seeing this sentence multiple times when working on this patch, but those were other `xxx_MAX_SIZE `macros.

[the remark this answers is not in the document]
Comment (yanesca) [reply include/psa/crypto_sizes.h:695]: Oh I see. Thank you @athoelke for the clarification! This means that the current value of the macro is correct.

@bensze01 Could you please extend the comment with Andrew's explanation?

File: include/psa/crypto_sizes.h [hunk 2]
```diff
@@ -657,4 +657,95 @@
...
+ * \param alg       An AEAD algorithm (\c PSA_ALG_XXX value such that #PSA_ALG_IS_AEAD(\p alg) is true).
+ *
+ * \return The default nonce size for the specified key type and algorithm.
+ *         If the key type or AEAD algorithm is not recognized,
+ *         or the parameters are incompatible, return 0.
+ *         An implementation can return either 0 or a correct size for a key type
+ *         and AEAD algorithm that it recognizes, but does not support.
+ */
+#define PSA_AEAD_NONCE_LENGTH(key_type, alg) \
+    (((key_type) == PSA_KEY_TYPE_AES || \
```
[the remark this answers is not in the document]
Comment (bensze01) [reply include/psa/crypto_sizes.h:686]: Done.

File: include/psa/crypto_sizes.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (bensze01) [reply include/psa/crypto_sizes.h:701]: @yanesca Done

**Review by yanesca — DISMISSED** [review]
LGTM.

File: include/psa/crypto_sizes.h [hunk 3]
```diff
@@ -657,4 +657,101 @@
...
+ *
+ * \return The default IV size for the specified key type and algorithm.
+ *         If the algorithm does not use an IV, return 0.
+ *         If the key type or cipher algorithm is not recognized,
+ *         or the parameters are incompatible, return 0.
+ *         An implementation can return either 0 or a correct size for a key type
+ *         and cipher algorithm that it recognizes, but does not support.
+ */
+#define PSA_CIPHER_IV_LENGTH(key_type, alg) \
+    ((key_type) == PSA_KEY_TYPE_AES && \
```
Comment (gilles-peskine-arm) [inline include/psa/crypto_sizes.h:730]: This definition shouldn't use the key type directly for block ciphers (`PSA_ALG_CIPHER_FROM_BLOCK_FLAG`). It should use the block size (`PSA_BLOCK_CIPHER_BLOCK_SIZE`).

**Review by gilles-peskine-arm — CHANGES_REQUESTED** [review]

File: include/psa/crypto_sizes.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (yanesca) [reply include/psa/crypto_sizes.h:686]: > On the contrary, I don't think PSA_ALG_IS_AEAD_ON_BLOCK_CIPHER should appear here.

I don't have strong opinion on this, I am happy either way.

> And isn't the maximum nonce size for CCM 13?

Yes, it is, but in the other discussion @athoelke said that these macros are about the default PSA sizes and not the maximum:
https://github.com/ARMmbed/mbedtls/pull/3787#discussion_r506188663
and we assumed that to be true.

File: include/psa/crypto_sizes.h [hunk 4]
```diff
@@ -657,4 +657,101 @@
...
+ * The IV that is generated as part of a call to #psa_cipher_encrypt() is always
+ * the default IV length for the algorithm.
+ *
+ * This macro can be used to allocate a buffer of sufficient size to
+ * store the IV output from #psa_cipher_generate_iv() when using
+ * a multi-part cipher operation.
+ *
+ * See also #PSA_CIPHER_IV_MAX_SIZE.
+ *
+ * \warning This function may call its arguments multiple times or
```
[the remark this answers is not in the document]
Comment (bensze01) [reply include/psa/crypto_sizes.h]: Done

File: include/psa/crypto_sizes.h (hunk 3 above)
Comment (bensze01) [reply include/psa/crypto_sizes.h]: Done

File: include/psa/crypto_sizes.h [hunk 5]
```diff
@@ -657,4 +657,101 @@
...
+ *
+ * \return The default nonce size for the specified key type and algorithm.
+ *         If the key type or AEAD algorithm is not recognized,
+ *         or the parameters are incompatible, return 0.
+ *         An implementation can return either 0 or a correct size for a key type
+ *         and AEAD algorithm that it recognizes, but does not support.
+ */
+#define PSA_AEAD_NONCE_LENGTH(key_type, alg) \
+    (((key_type) == PSA_KEY_TYPE_AES || \
+      (key_type) == PSA_KEY_TYPE_CAMELLIA) && \
```
[the remark this answers is not in the document]
Comment (bensze01) [reply include/psa/crypto_sizes.h]: I've changed the check to use the block size. On a side note, [the PSA Cryto API doesn't define ARIA as a supported key type](https://armmbed.github.io/mbed-crypto/html/api/keys/types.html).

File: include/psa/crypto_sizes.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (bensze01) [reply include/psa/crypto_sizes.h]: @gilles-peskine-arm I've changed the check back to the original version, checking for CCM and GCM explicitly. I've left the size for CCM at 12, as per @yanesca 's comment.

File: include/psa/crypto_sizes.h [hunk 6]
```diff
@@ -657,4 +657,101 @@
...
+          PSA_ALG_IS_AEAD_ON_BLOCK_CIPHER(alg) ? 12 : \
+      (key_type) == PSA_KEY_TYPE_CHACHA20 && \
```
[the remark this answers is not in the document]
Comment (bensze01) [reply include/psa/crypto_sizes.h]: Yes, but the API requires us to return 0 if the key type and algorithm are incompatible, so we still ought to check if the correct key type was provided, no?

File: include/psa/crypto_sizes.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (gilles-peskine-arm) [reply include/psa/crypto_sizes.h]: Oops. I thought PSA declared all key types that Mbed TLS supports except for a few obsolete block ciphers. We'll add ARIA along with a few others in a 1.0.x update of the API document.

File: include/psa/crypto_sizes.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (gilles-peskine-arm) [reply include/psa/crypto_sizes.h]: Ah, yes, good point.

**Review by yanesca — APPROVED** [review]
Looks good to me.

**Review by gilles-peskine-arm — APPROVED** [review]

**Pull request merged**